libdnf/rpm/repo_sack.cpp

```diff
@@ -19,22 +19,58 @@ along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
 
 #include "libdnf/rpm/repo_sack.hpp"
 
+#include "repo_impl.hpp"
+
 #include "libdnf/base/base.hpp"
 #include "libdnf/conf/config_parser.hpp"
 #include "libdnf/conf/option_bool.hpp"
+#include "libdnf/rpm/solv_sack_impl.hpp"
 
 #include <fmt/format.h>
 
+extern "C" {
+#include <solv/repo.h>
+#include <solv/solv_xfopen.h>
+#include <solv/testcase.h>
+}
+
 #include <filesystem>
 
+
+using LibsolvRepo = ::Repo;
+
+
 namespace libdnf::rpm {
 
+
 RepoWeakPtr RepoSack::new_repo(const std::string & id) {
     // TODO(jrohel): Test repo exists
     auto repo = std::make_unique<Repo>(id, *base, Repo::Type::AVAILABLE);
     return add_item_with_return(std::move(repo));
 }
 
+// Deleter for std::unique_ptr<LibsolvRepo>
+static void libsolv_repo_free(LibsolvRepo * libsolv_repo) {
+    repo_free(libsolv_repo, 1);
+}
+
+
+RepoWeakPtr RepoSack::new_repo_from_libsolv_testcase(const std::string & repoid, const std::string & path) {
+    std::unique_ptr<std::FILE, decltype(&std::fclose)> testcase_file(solv_xfopen(path.c_str(), "r"), &std::fclose);
+    if (!testcase_file) {
+        throw std::runtime_error(fmt::format("Unable to open libsolv testcase file: {}", path));
+    }
+
+    auto repo = new_repo(repoid);
+    Pool * pool = base->get_rpm_solv_sack().p_impl->get_pool();
```

Comment:
I didn't investigate this much, but... it seems arbitrary and a bit awkward to pull the `pool` from `base->get_rpm_solv_sack().p_impl` and also seems to be the reason why you had to add the class as a friend maybe? We should do this in a better way. Feel free to defer this to a later iteration...

Reply:
This needs to be made more straightforward in the next iterations,
but it's not arbitrary. Everything is in a single pool (repos, packages) and that owned by SolvSack (that contains Package objects) now.